### PR TITLE
Rewrite handling of fallthrough successor reification

### PR DESCRIPTION
**Pull request by travitch**
The fallthrough successor reification accounts for cases where control flow
falls through the end of one basic block and proceeds at the beginning of the
immediate successor block.  This is common in e.g., conditional branches.

The new code is simpler and makes it clearer how the logic works.  More
importantly, the successor computations occur earlier in the process when we
have accurate concrete blocks available.  This will be very useful for some
upcoming changes.

Additionally, the revised handling allows us to decouple fallthrough reification
from the last instruction of the block, allowing instrumentation to be inserted
e.g., after calls that end a block, but before the fallthrough control flow is
inserted.

### CODE REVIEW

**Comment by travitch** [comment]
NOTE: This isn't ready to merge yet - the architecture-specific backends need to be updated

@langston-barrett @m10f I think that this change should let you add instrumentation after function calls.

@dmwit I'd appreciate your eyeballs on this change... I think it is conceptually pretty simple, but it touched a lot of things.  Can you give it a look and see if there is any obvious design problem?  I haven't had a chance to test it yet, since I need to update the x86 and PowerPC backends.

**Comment by langston-barrett** [comment]
@travitch First impression: I love replacing `error` with `panic`! It makes it much more clear that it's a programming error, not e.g. malformed input or user error.

File: renovate/src/Renovate/ISA.hs [hunk 1]
```diff
@@ -118,7 +118,22 @@ data ISA arch = ISA
     -- New code blocks will be laid out in virtual address space within this
     -- many bytes of the original code blocks, so that the two can jump to each
     -- other as necessary.
-  , isaModifyJumpTarget :: ConcreteAddress arch -> ConcreteFallthrough arch () -> Maybe [Instruction arch ()]
+  , isaModifyJumpTarget :: ConcreteAddress arch -> Instruction arch () -> Maybe (DLN.NonEmpty (Instruction arch ()))
+  -- , isaModifyJumpTarget :: ConcreteAddress arch -> ConcreteFallthrough arch () -> Maybe [Instruction arch ()]
```
Comment (langston-barrett) [inline renovate/src/Renovate/ISA.hs]: Delete?

**Comment by travitch** [comment]
I'm a bit torn on panic.  I agree that the clarity there is nice, but on the other hand renovate is a library and errors should probably be recoverable.

File: renovate/src/Renovate/ISA.hs [hunk 2]
```diff
@@ -118,7 +118,22 @@ data ISA arch = ISA
...
+    -- ^ Modify the given jump instruction, accounting for any fallthrough behavior specified
```
Comment (dmwit) [inline renovate/src/Renovate/ISA.hs]: Are there now two haddock comments for this?

File: renovate/src/Renovate/Redirect/LayoutBlocks/Compact.hs [hunk 3]
```diff
@@ -77,11 +77,25 @@ compactLayout startAddr strat blocks0 injectedCode cfgs = do
...
+  -- concretize) to add explicit fallthroughs for the *last* block in these
+  -- groups.  We can remove all of the other fallthroughs here.
+  --
+  -- We want to do this as early as possible because the number of fallthroughs
+  -- affect the ultimate sizes of blocks, and we need to be able to accurately
+  -- compute block sizes to do layout.
+  --
+  -- It is important to do this before we compute any block sizes so that things
+  -- are all internally consistent.
+  blockChunks <- removeUnneededFallthroughs blockChunks0
```
Comment (dmwit) [inline renovate/src/Renovate/Redirect/LayoutBlocks/Compact.hs]: Love this comment. Was watching for where this behavior would happen, and this made it super easy to spot.

File: renovate/src/Renovate/Redirect/Symbolize.hs [hunk 4]
```diff
@@ -81,8 +82,11 @@ symbolizeJumps isa mem symAddrMap (cb, symAddr) =
   case DLN.nonEmpty (concat insns) of
     Nothing -> error ("Created empty block while symbolizing block at: " ++ show (concreteBlockAddress cb))
     Just insnList ->
-      (cb, symbolicBlock (concreteBlockAddress cb) symAddr insnList)
+      let msucc = lookupSymAddr =<< explicitFallthroughSuccessor fallthrough
+      in (cb, symbolicBlock cb symAddr insnList msucc)
```
Comment (dmwit) [inline renovate/src/Renovate/Redirect/Symbolize.hs]: Hm, is this really the right behavior? The way I read this, it says that if there's an explicit fallthrough address in `fallthrough`, but we can't find that address in our `symAddrMap`, we claim that no fallthrough is needed. Maybe we should report an error of some kind instead?

Comment (travitch) [reply renovate/src/Renovate/Redirect/Symbolize.hs]: Good catch - you are right

File: renovate/src/Renovate/ISA.hs (hunk 2 above)
Comment (dmwit) [reply renovate/src/Renovate/ISA.hs]: Ah, fixed in a later commit. Sorry for the noise.

**Comment by dmwit** [comment]
Top level nothing jumps out as super problematic to me. The two tricky bits of logic I was watching for was avoiding fallthroughs when blocks are relocated together and when they end in unconditional jumps, and it looks like that logic got moved appropriately in this commit, so that much looks good to me.

**Comment by travitch** [comment]
Thanks @dmwit.  After running the tests and wresting with it a bit, I got this working.  As expected, a few subtle errors popped up.  You might find the fixes amusing and/or interesting.

@langston-barrett and @m10f: I'll merge this to master shortly (Monday morning at the latest).  It accommodates your use case and will allow you to insert instrumentation after a call.

**Pull request merged**